// @sanity/block-tools/lib/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SLATE_DEFAULT_BLOCK = exports.HTML_SPAN_TAGS = exports.HTML_MISC_TAGS = exports.HTML_LIST_ITEM_TAGS = exports.HTML_LIST_CONTAINER_TAGS = exports.HTML_HEADER_TAGS = exports.HTML_DECORATOR_TAGS = exports.HTML_BLOCK_TAGS = exports.ELEMENT_MAP = exports.DEFAULT_SUPPORTED_STYLES = exports.DEFAULT_SUPPORTED_DECORATORS = exports.DEFAULT_SUPPORTED_ANNOTATIONS = exports.DEFAULT_SPAN = exports.DEFAULT_BLOCK = exports.BLOCK_DEFAULT_STYLE = void 0;
var _uniq2 = _interopRequireDefault(require("lodash/uniq"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var BLOCK_DEFAULT_STYLE = exports.BLOCK_DEFAULT_STYLE = 'normal';

// Slate block
var SLATE_DEFAULT_BLOCK = exports.SLATE_DEFAULT_BLOCK = Object.freeze({
  object: 'block',
  type: 'contentBlock',
  data: {
    style: BLOCK_DEFAULT_STYLE
  }
});
var DEFAULT_BLOCK = exports.DEFAULT_BLOCK = Object.freeze({
  _type: 'block',
  markDefs: [],
  style: BLOCK_DEFAULT_STYLE
});
var DEFAULT_SPAN = exports.DEFAULT_SPAN = Object.freeze({
  _type: 'span',
  marks: []
});
var HTML_BLOCK_TAGS = exports.HTML_BLOCK_TAGS = {
  p: DEFAULT_BLOCK,
  blockquote: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'blockquote'
  })
};
var HTML_SPAN_TAGS = exports.HTML_SPAN_TAGS = {
  span: {
    object: 'text'
  }
};
var HTML_LIST_CONTAINER_TAGS = exports.HTML_LIST_CONTAINER_TAGS = {
  ol: {
    object: null
  },
  ul: {
    object: null
  }
};
var HTML_HEADER_TAGS = exports.HTML_HEADER_TAGS = {
  h1: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h1'
  }),
  h2: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h2'
  }),
  h3: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h3'
  }),
  h4: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h4'
  }),
  h5: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h5'
  }),
  h6: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: 'h6'
  })
};
var HTML_MISC_TAGS = exports.HTML_MISC_TAGS = {
  br: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: BLOCK_DEFAULT_STYLE
  })
};
var HTML_DECORATOR_TAGS = exports.HTML_DECORATOR_TAGS = {
  b: 'strong',
  strong: 'strong',
  i: 'em',
  em: 'em',
  u: 'underline',
  s: 'strike-through',
  strike: 'strike-through',
  del: 'strike-through',
  code: 'code'
};
var HTML_LIST_ITEM_TAGS = exports.HTML_LIST_ITEM_TAGS = {
  li: _objectSpread(_objectSpread({}, DEFAULT_BLOCK), {}, {
    style: BLOCK_DEFAULT_STYLE,
    level: 1,
    listItem: 'bullet'
  })
};
var ELEMENT_MAP = exports.ELEMENT_MAP = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, HTML_BLOCK_TAGS), HTML_SPAN_TAGS), HTML_LIST_CONTAINER_TAGS), HTML_LIST_ITEM_TAGS), HTML_HEADER_TAGS), HTML_MISC_TAGS);
var DEFAULT_SUPPORTED_STYLES = exports.DEFAULT_SUPPORTED_STYLES = (0, _uniq2.default)(Object.keys(ELEMENT_MAP).filter(tag => ELEMENT_MAP[tag].style).map(tag => ELEMENT_MAP[tag].style));
var DEFAULT_SUPPORTED_DECORATORS = exports.DEFAULT_SUPPORTED_DECORATORS = (0, _uniq2.default)(Object.keys(HTML_DECORATOR_TAGS).map(tag => HTML_DECORATOR_TAGS[tag]));
var DEFAULT_SUPPORTED_ANNOTATIONS = exports.DEFAULT_SUPPORTED_ANNOTATIONS = ['link'];
