// @stackbit/cms-core/src/types/custom-actions.ts
import {
    CustomActionGlobal,
    CustomActionBulk,
    CustomActionDocument,
    CustomActionObjectModel,
    CustomActionObjectField,
    CustomActionField,
    CustomActionState,
    CustomActionInputField,
    CustomActionPermissions
} from '@stackbit/types';
import { User } from './index';

export type CustomActionRunStateMap = Record<string, CustomActionRunState>;
export type CustomActionRunState = {
    // specifies if the action is current running in this container instance
    runningHandler?: boolean;
    // the action state returned by the recent 'run' invocation
    lastResultState?: CustomActionState;
};

export type ExtendedCustomAction =
    | ExtendedCustomActionGlobal
    | ExtendedCustomActionBulk
    | ExtendedCustomActionDocument
    | ExtendedCustomActionObjectModel
    | ExtendedCustomActionObjectField
    | ExtendedCustomActionField;

export type ExtendedCustomActionGlobal = CustomActionGlobal & ContentStoreCustomActionCommonProps;
export type ExtendedCustomActionBulk = CustomActionBulk & ContentStoreCustomActionCommonProps;
export type ExtendedCustomActionDocument = CustomActionDocument &
    ContentStoreCustomActionCommonProps & {
        type: 'document';
        documentSpec: APICustomActionDocumentSpecifier;
    };
export type ExtendedCustomActionObjectModel = Omit<CustomActionObjectModel, 'type'> &
    ContentStoreCustomActionCommonProps & {
        type: 'objectModel';
        documentSpec: APICustomActionDocumentSpecifier;
        fieldPath: (string | number)[];
    };
export type ExtendedCustomActionObjectField = Omit<CustomActionObjectField, 'type'> &
    ContentStoreCustomActionCommonProps & {
        type: 'objectField';
        documentSpec: APICustomActionDocumentSpecifier;
        fieldPath: (string | number)[];
    };
export type ExtendedCustomActionField = CustomActionField &
    ContentStoreCustomActionCommonProps & {
        type: 'field';
        documentSpec: APICustomActionDocumentSpecifier;
        fieldPath: (string | number)[];
    };
export type ContentStoreCustomActionCommonProps = {
    type: string;
    actionId: string;
    // override optional label to required one
    label: string;
    // specifies if the action is current running in this container instance
    runningHandler?: boolean;
    // the action state returned by the recent 'run' invocation
    lastResultState?: CustomActionState;
};

export type CustomActionType = 'global' | 'bulk' | 'document' | 'object' | 'field';

export type APICustomAction = APICustomActionGlobal | APICustomActionBulk | APICustomActionDocument | APICustomActionObject | APICustomActionField;

export interface APICustomActionGlobal extends APICustomActionCommonProps {
    type: 'global';
}

export interface APICustomActionBulk extends APICustomActionCommonProps {
    type: 'bulk';
}

export interface APICustomActionDocument extends APICustomActionCommonProps {
    type: 'document';
}

export interface APICustomActionObject extends APICustomActionCommonProps {
    type: 'object';
}

export interface APICustomActionField extends APICustomActionCommonProps {
    type: 'field';
}

export interface APICustomActionCommonProps {
    type: CustomActionType;
    name: string;
    label: string;
    icon?: string;
    inputFields?: CustomActionInputField[];
    actionId: string;
    state: CustomActionState;
    hidden?: boolean;
    permissions?: CustomActionPermissions;
}

export type APIRunCustomActionRequest =
    | APIRunCustomActionRequestGlobal
    | APIRunCustomActionRequestBulk
    | APIRunCustomActionRequestDocument
    | APIRunCustomActionRequestObject
    | APIRunCustomActionRequestField;

export interface APIRunCustomActionRequestGlobal extends APIRunCustomActionRequestCommonParams {
    actionType: 'global';
}

export interface APIRunCustomActionRequestBulk extends APIRunCustomActionRequestCommonParams {
    actionType: 'bulk';
    documents: APICustomActionDocumentSpecifier[];
}

export interface APIRunCustomActionRequestDocument extends APIRunCustomActionRequestCommonParams {
    actionType: 'document';
}

export interface APIRunCustomActionRequestObject extends APIRunCustomActionRequestCommonParams {
    actionType: 'object';
}

export interface APIRunCustomActionRequestField extends APIRunCustomActionRequestCommonParams {
    actionType: 'field';
}

interface APIRunCustomActionRequestCommonParams {
    actionType: string;
    actionId: string;
    actionName: string;
    locale?: string;
    user?: User;
    pageUrl?: string;
    currentPageDocument?: APICustomActionDocumentSpecifier; // the document id of the current page
    inputData?: Record<string, any>;
}

export interface APIGetCustomActionRequest {
    customActionIds: string[];
    locale?: string;
    user?: User;
    pageUrl?: string;
    currentPageDocument?: APICustomActionDocumentSpecifier; // the document id of the current page
}

export interface APICustomActionDocumentSpecifier {
    srcType: string;
    srcProjectId: string;
    srcDocumentId: string;
}

export interface CustomActionStateChange {
    actionType: CustomActionType;
    actionId: string;
    actionName: string;
    state: CustomActionState;
    success?: string;
    error?: string;
    result?: unknown;
}
