// @stackbit/cms-contentful/dist/contentful-consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CONTENTFUL_BUILT_IN_IMAGE_SOURCES = exports.CONTENTFUL_APRIMO_APP = exports.CONTENTFUL_BYNDER_APP = exports.CONTENTFUL_CLOUDINARY_APP = exports.CONTENTFUL_MARKS_TYPES_MAP = exports.CONTENTFUL_NODE_TYPES_MAP = void 0;
const rich_text_types_1 = require("@contentful/rich-text-types");
const { RICH_TEXT_NODE_TYPES, RICH_TEXT_MARK_TYPES } = require('@stackbit/types');
exports.CONTENTFUL_NODE_TYPES_MAP = {
    [rich_text_types_1.BLOCKS.PARAGRAPH]: RICH_TEXT_NODE_TYPES.PARAGRAPH,
    [rich_text_types_1.BLOCKS.QUOTE]: RICH_TEXT_NODE_TYPES.BLOCKQUOTE,
    [rich_text_types_1.BLOCKS.UL_LIST]: RICH_TEXT_NODE_TYPES.UL,
    [rich_text_types_1.BLOCKS.OL_LIST]: RICH_TEXT_NODE_TYPES.OL,
    [rich_text_types_1.BLOCKS.LIST_ITEM]: RICH_TEXT_NODE_TYPES.LIST_ITEM,
    [rich_text_types_1.BLOCKS.HEADING_1]: RICH_TEXT_NODE_TYPES.HEADING_1,
    [rich_text_types_1.BLOCKS.HEADING_2]: RICH_TEXT_NODE_TYPES.HEADING_2,
    [rich_text_types_1.BLOCKS.HEADING_3]: RICH_TEXT_NODE_TYPES.HEADING_3,
    [rich_text_types_1.BLOCKS.HEADING_4]: RICH_TEXT_NODE_TYPES.HEADING_4,
    [rich_text_types_1.BLOCKS.HEADING_5]: RICH_TEXT_NODE_TYPES.HEADING_5,
    [rich_text_types_1.BLOCKS.HEADING_6]: RICH_TEXT_NODE_TYPES.HEADING_6,
    [rich_text_types_1.BLOCKS.HR]: RICH_TEXT_NODE_TYPES.HR,
    [rich_text_types_1.INLINES.HYPERLINK]: RICH_TEXT_NODE_TYPES.LINK,
    [rich_text_types_1.BLOCKS.EMBEDDED_ASSET]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [rich_text_types_1.BLOCKS.EMBEDDED_ENTRY]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [rich_text_types_1.INLINES.EMBEDDED_ENTRY]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [rich_text_types_1.INLINES.ASSET_HYPERLINK]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [rich_text_types_1.INLINES.ENTRY_HYPERLINK]: RICH_TEXT_NODE_TYPES.REFERENCE
};
exports.CONTENTFUL_MARKS_TYPES_MAP = {
    [rich_text_types_1.MARKS.BOLD]: RICH_TEXT_MARK_TYPES.BOLD,
    [rich_text_types_1.MARKS.ITALIC]: RICH_TEXT_MARK_TYPES.ITALIC,
    [rich_text_types_1.MARKS.UNDERLINE]: RICH_TEXT_MARK_TYPES.UNDERLINE,
    [rich_text_types_1.MARKS.CODE]: RICH_TEXT_MARK_TYPES.CODE
};
exports.CONTENTFUL_CLOUDINARY_APP = 'zjcnWgBknf9zB7IM9HZjE';
exports.CONTENTFUL_BYNDER_APP = '5KySdUzG7OWuCE2V3fgtIa';
exports.CONTENTFUL_APRIMO_APP = 'RkD9DMLgnqahWzcYyQ0RI';
exports.CONTENTFUL_BUILT_IN_IMAGE_SOURCES = ['cloudinary', 'bynder', 'aprimo'];
