// @stackbit/cms-contentful/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchAndConvertSchema = exports.localizeModels = exports.localizeFieldData = exports.ContentfulEncoderDelegate = void 0;
const contentful_api_client_1 = require("./contentful-api-client");
const contentful_schema_converter_1 = require("./contentful-schema-converter");
__exportStar(require("./contentful-content-source"), exports);
__exportStar(require("./sync-poller"), exports);
var contentful_encoder_delegate_1 = require("./contentful-encoder-delegate");
Object.defineProperty(exports, "ContentfulEncoderDelegate", { enumerable: true, get: function () { return __importDefault(contentful_encoder_delegate_1).default; } });
var localize_field_data_1 = require("./localize-field-data");
Object.defineProperty(exports, "localizeFieldData", { enumerable: true, get: function () { return localize_field_data_1.localizeFieldData; } });
var localize_models_1 = require("./localize-models");
Object.defineProperty(exports, "localizeModels", { enumerable: true, get: function () { return localize_models_1.localizeModels; } });
async function fetchAndConvertSchema(options) {
    const client = (0, contentful_api_client_1.createPlainApiClient)({
        accessToken: options.accessToken ?? process.env.CONTENTFUL_ACCESS_TOKEN,
        spaceId: options.spaceId ?? process.env.CONTENTFUL_SPACE_ID,
        environment: options.environment ?? process.env.CONTENTFUL_ENVIRONMENT ?? 'master'
    });
    const contentTypes = await (0, contentful_api_client_1.fetchAllContentTypes)(client);
    const editorInterfaces = await (0, contentful_api_client_1.fetchAllEditorInterfaces)(client);
    const locales = await (0, contentful_api_client_1.fetchAllLocales)(client);
    const defaultLocaleCode = locales.find((locale) => locale.default)?.code;
    return (0, contentful_schema_converter_1.convertSchema)({
        contentTypes,
        editorInterfaces,
        defaultLocaleCode: defaultLocaleCode ?? 'en-US',
        cloudinaryImagesAsList: false,
        bynderImagesAsList: false // not supported in non CSI at all
    });
}
exports.fetchAndConvertSchema = fetchAndConvertSchema;
