// @stackbit/cms-contentful/src/contentful-consts.ts
import { BLOCKS, INLINES, MARKS } from '@contentful/rich-text-types';
const { RICH_TEXT_NODE_TYPES, RICH_TEXT_MARK_TYPES } = require('@stackbit/types');

export const CONTENTFUL_NODE_TYPES_MAP = {
    [BLOCKS.PARAGRAPH]: RICH_TEXT_NODE_TYPES.PARAGRAPH,
    [BLOCKS.QUOTE]: RICH_TEXT_NODE_TYPES.BLOCKQUOTE,
    [BLOCKS.UL_LIST]: RICH_TEXT_NODE_TYPES.UL,
    [BLOCKS.OL_LIST]: RICH_TEXT_NODE_TYPES.OL,
    [BLOCKS.LIST_ITEM]: RICH_TEXT_NODE_TYPES.LIST_ITEM,
    [BLOCKS.HEADING_1]: RICH_TEXT_NODE_TYPES.HEADING_1,
    [BLOCKS.HEADING_2]: RICH_TEXT_NODE_TYPES.HEADING_2,
    [BLOCKS.HEADING_3]: RICH_TEXT_NODE_TYPES.HEADING_3,
    [BLOCKS.HEADING_4]: RICH_TEXT_NODE_TYPES.HEADING_4,
    [BLOCKS.HEADING_5]: RICH_TEXT_NODE_TYPES.HEADING_5,
    [BLOCKS.HEADING_6]: RICH_TEXT_NODE_TYPES.HEADING_6,
    [BLOCKS.HR]: RICH_TEXT_NODE_TYPES.HR,
    [INLINES.HYPERLINK]: RICH_TEXT_NODE_TYPES.LINK,
    [BLOCKS.EMBEDDED_ASSET]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [BLOCKS.EMBEDDED_ENTRY]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [INLINES.EMBEDDED_ENTRY]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [INLINES.ASSET_HYPERLINK]: RICH_TEXT_NODE_TYPES.REFERENCE,
    [INLINES.ENTRY_HYPERLINK]: RICH_TEXT_NODE_TYPES.REFERENCE
};

export const CONTENTFUL_MARKS_TYPES_MAP = {
    [MARKS.BOLD]: RICH_TEXT_MARK_TYPES.BOLD,
    [MARKS.ITALIC]: RICH_TEXT_MARK_TYPES.ITALIC,
    [MARKS.UNDERLINE]: RICH_TEXT_MARK_TYPES.UNDERLINE,
    [MARKS.CODE]: RICH_TEXT_MARK_TYPES.CODE
};

export const CONTENTFUL_CLOUDINARY_APP = 'zjcnWgBknf9zB7IM9HZjE';
export const CONTENTFUL_BYNDER_APP = '5KySdUzG7OWuCE2V3fgtIa';
export const CONTENTFUL_APRIMO_APP = 'RkD9DMLgnqahWzcYyQ0RI';

export const CONTENTFUL_BUILT_IN_IMAGE_SOURCES = ['cloudinary', 'bynder', 'aprimo'];
