// @stackbit/cms-core/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encodeData = exports.ContentStoreTypes = exports.consts = exports.getObjectId = exports.getDocumentFieldForLocale = exports.referencesUtils = exports.searchUtils = exports.utils = exports.annotator = exports.stackbit = void 0;
exports.stackbit = __importStar(require("./stackbit"));
exports.annotator = __importStar(require("./annotator"));
exports.utils = __importStar(require("./utils"));
exports.searchUtils = __importStar(require("./utils/search-utils"));
exports.referencesUtils = __importStar(require("./utils/references-utils"));
__exportStar(require("./utils/field-path-utils"), exports);
var content_store_utils_1 = require("./content-store-utils");
Object.defineProperty(exports, "getDocumentFieldForLocale", { enumerable: true, get: function () { return content_store_utils_1.getDocumentFieldForLocale; } });
Object.defineProperty(exports, "getObjectId", { enumerable: true, get: function () { return content_store_utils_1.getObjectId; } });
exports.consts = __importStar(require("./consts"));
__exportStar(require("./common/common-schema"), exports);
__exportStar(require("./common/common-types"), exports);
__exportStar(require("./content-store"), exports);
exports.ContentStoreTypes = __importStar(require("./types"));
var encoder_1 = require("./encoder");
Object.defineProperty(exports, "encodeData", { enumerable: true, get: function () { return __importDefault(encoder_1).default; } });
__exportStar(require("./encoder"), exports);
__exportStar(require("./services"), exports);
