// @stackbit/cms-core/dist/utils/csi-to-api-docs-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapDocumentVersionsToApiDocumentVersions = void 0;
const csi_to_store_docs_converter_1 = require("./csi-to-store-docs-converter");
const store_to_api_docs_converter_1 = require("./store-to-api-docs-converter");
const store_to_api_v2_docs_converter_1 = require("./store-to-api-v2-docs-converter");
function mapDocumentVersionsToApiDocumentVersions({ versions, contentSourceData, locale, assetSources, createConfigDelegate, contentSourceDataById, modelMap }) {
    return versions.map((version) => {
        if (!version.document) {
            return version;
        }
        const delegate = createConfigDelegate();
        const contentStoreDocuments = (0, csi_to_store_docs_converter_1.mapCSIDocumentsToStoreDocuments)({
            csiDocuments: [version.document],
            contentSourceInstance: contentSourceData.instance,
            defaultLocaleCode: contentSourceData.defaultLocaleCode,
            modelMap,
            createConfigDelegate,
            assetSources,
            logger: delegate.getLogger()
        });
        const [object] = (0, store_to_api_docs_converter_1.mapDocumentsToLocalizedApiObjects)({
            documents: contentStoreDocuments,
            locale,
            delegate
        });
        const [apiDocument] = (0, store_to_api_v2_docs_converter_1.mapDocumentsToApiDocuments)({
            documents: contentStoreDocuments,
            contentSourceDataById,
            delegate
        });
        return {
            ...version,
            object,
            apiDocument
        };
    });
}
exports.mapDocumentVersionsToApiDocumentVersions = mapDocumentVersionsToApiDocumentVersions;
