// @stackbit/cms-core/dist/utils/meta-fields-filters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.META_FIELDS = void 0;
const search_utils_1 = require("./search-utils");
exports.META_FIELDS = {
    createdAt: ({ filter, document }) => ({ field: { type: 'date', value: document.createdAt } }),
    updatedAt: ({ filter, document }) => ({ field: { type: 'date', value: document.updatedAt } }),
    status: ({ filter, document }) => ({ field: { type: 'enum', value: document.status } }),
    hasScheduledActions: ({ filter, document, context }) => ({
        field: { type: 'boolean', value: !!context.activeScheduledActionsByDocumentId[document.srcObjectId]?.length }
    }),
    scheduledActionDate: ({ filter, document, context }) => {
        return {
            result: context.activeScheduledActionsByDocumentId[document.srcObjectId]?.some((scheduledAction) => {
                const field = { type: 'date', value: scheduledAction.executeAt };
                return (0, search_utils_1.isDateFieldMatches)({ field, filter, locale: context.locale });
            }) ?? false
        };
    }
};
