// @stackbit/cms-core/dist/utils/store-to-csi-docs-converter.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mapStoreFieldToCSIField = exports.mapStoreDocumentToCSIDocumentWithSource = exports.mapStoreDocumentsToCSIDocumentsWithSource = void 0;
const lodash_1 = __importDefault(require("lodash"));
const utils_1 = require("@stackbit/utils");
function mapStoreDocumentsToCSIDocumentsWithSource({ documents, csiDocumentMap }) {
    return documents.map((document) => mapStoreDocumentToCSIDocumentWithSource({
        document,
        csiDocument: csiDocumentMap[document.srcObjectId]
    }));
}
exports.mapStoreDocumentsToCSIDocumentsWithSource = mapStoreDocumentsToCSIDocumentsWithSource;
/**
 * This method converts documents stored in an internal ContentStoreTypes.Document
 * format to documents in a public CSITypes.DocumentWithSource format. These
 * documents are then passed to document hook and sitemap functions defined in
 * stackbit.config.ts.
 *
 * Most of the properties, including fields and their types, are taken from the
 * internal documents that contain changes made by mapDocuments, mapModels and
 * modelExtensions functions.
 *
 * The document's context is taken from the CSI Document because context is not
 * included in the internal documents.
 */
function mapStoreDocumentToCSIDocumentWithSource({ document, csiDocument }) {
    return {
        type: document.type,
        id: document.srcObjectId,
        srcType: document.srcType,
        srcProjectId: document.srcProjectId,
        manageUrl: document.srcObjectUrl,
        modelName: document.srcModelName,
        status: document.status,
        createdAt: document.createdAt,
        createdBy: document.createdBy,
        updatedAt: document.updatedAt,
        updatedBy: document.updatedBy,
        locale: document.locale,
        context: csiDocument.context,
        hidden: document.hidden,
        fields: mapStoreFieldsToCSIFields(document.fields)
    };
}
exports.mapStoreDocumentToCSIDocumentWithSource = mapStoreDocumentToCSIDocumentWithSource;
function mapStoreFieldsToCSIFields(documentFields) {
    return lodash_1.default.reduce(documentFields, (accum, documentField, fieldName) => {
        const csiField = mapStoreFieldToCSIField(documentField);
        if (csiField) {
            accum[fieldName] = csiField;
        }
        return accum;
    }, {});
}
function mapStoreFieldToCSIField(documentField) {
    switch (documentField.type) {
        case 'string':
        case 'text':
        case 'html':
        case 'slug':
        case 'url':
        case 'color':
        case 'boolean':
        case 'number':
        case 'date':
        case 'datetime':
        case 'enum':
        case 'file':
        case 'json':
        case 'style':
        case 'markdown':
        case 'richText': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: documentField.type,
                    localized: true,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        value: locale.value
                    }))
                };
            }
            if (typeof documentField.value === 'undefined' || ('isUnset' in documentField && documentField.isUnset)) {
                return;
            }
            return {
                type: documentField.type,
                value: documentField.value
            };
        }
        case 'image': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return (0, utils_1.omitByNil)({
                    type: 'image',
                    localized: true,
                    source: documentField.source,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => (0, utils_1.omitByNil)({
                        locale: locale.locale,
                        sourceData: locale.sourceData,
                        ...(locale.sourceData ? null : { fields: mapStoreFieldsToCSIFields(locale.fields) })
                    }))
                });
            }
            if (documentField.isUnset) {
                return;
            }
            return (0, utils_1.omitByNil)({
                type: 'image',
                source: documentField.source,
                sourceData: documentField.sourceData,
                ...(documentField.sourceData ? null : { fields: mapStoreFieldsToCSIFields(documentField.fields) })
            });
        }
        case 'object': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: 'object',
                    localized: true,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        fields: mapStoreFieldsToCSIFields(locale.fields)
                    }))
                };
            }
            if (documentField.isUnset) {
                return;
            }
            return {
                type: 'object',
                fields: mapStoreFieldsToCSIFields(documentField.fields)
            };
        }
        case 'model': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: 'model',
                    localized: true,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        modelName: locale.srcModelName,
                        fields: mapStoreFieldsToCSIFields(locale.fields)
                    }))
                };
            }
            if (documentField.isUnset) {
                return;
            }
            return {
                type: 'model',
                modelName: documentField.srcModelName,
                fields: mapStoreFieldsToCSIFields(documentField.fields)
            };
        }
        case 'reference': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: 'reference',
                    localized: true,
                    refType: documentField.refType,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        refId: locale.refId
                    }))
                };
            }
            if (documentField.isUnset) {
                return;
            }
            return {
                type: 'reference',
                refType: documentField.refType,
                refId: documentField.refId
            };
        }
        case 'cross-reference': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: 'json',
                    localized: true,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        value: {
                            refId: locale.refId,
                            refSrcType: locale.refSrcType,
                            srcProjectId: locale.refProjectId
                        }
                    }))
                };
            }
            if (documentField.isUnset) {
                return;
            }
            return {
                type: 'json',
                value: {
                    refId: documentField.refId,
                    refSrcType: documentField.refSrcType,
                    srcProjectId: documentField.refProjectId
                }
            };
        }
        case 'list': {
            if (documentField.localized) {
                if (lodash_1.default.isEmpty(documentField.locales)) {
                    return;
                }
                return {
                    type: 'list',
                    localized: true,
                    locales: lodash_1.default.mapValues(documentField.locales, (locale) => ({
                        locale: locale.locale,
                        items: locale.items.reduce((accum, item) => {
                            const csiItem = mapStoreFieldToCSIField(item);
                            // shouldn't happen because all list item fields are non-localized by definition,
                            // but just in case and for the sake of typescript
                            if (!csiItem) {
                                return accum;
                            }
                            return accum.concat(csiItem);
                        }, [])
                    }))
                };
            }
            return {
                type: 'list',
                items: documentField.items.reduce((accum, item) => {
                    const csiItem = mapStoreFieldToCSIField(item);
                    // shouldn't happen because all list item fields are non-localized by definition,
                    // but just in case and for the sake of typescript
                    if (!csiItem) {
                        return accum;
                    }
                    return accum.concat(csiItem);
                }, [])
            };
        }
        default: {
            const _exhaustiveCheck = documentField;
            return _exhaustiveCheck;
        }
    }
}
exports.mapStoreFieldToCSIField = mapStoreFieldToCSIField;
