// @stackbit/sdk/src/consts.ts
import _ from 'lodash';
import type { SSGRunOptions } from './config/config-types';

const excludedMarkdownFiles = ['LICENSE.md', '**/README.md', 'README.theme.md', 'CONTRIBUTING.md', 'CHANGELOG.md', 'CODE_OF_CONDUCT.md'];

export const MARKDOWN_FILE_EXTENSIONS = ['md', 'mdx', 'markdown'];
export const DATA_FILE_EXTENSIONS = ['yml', 'yaml', 'json', 'toml'];
export const SUPPORTED_FILE_EXTENSIONS = [...MARKDOWN_FILE_EXTENSIONS, ...DATA_FILE_EXTENSIONS];
export const ASSET_FILE_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'svg', 'ico', 'webp', 'avif', 'tiff', 'bmp'];
export const EXCLUDED_MARKDOWN_FILES = _.concat(excludedMarkdownFiles, excludedMarkdownFiles.map(_.toLower));
export const EXCLUDED_DATA_FILES = ['stackbit.yaml', 'netlify.toml', 'theme.toml', '**/package.json', '**/package-lock.json', '**/yarn-lock.json'];
export const EXCLUDED_COMMON_FILES = ['**/node_modules/**', '**/.git/**', '.idea/**', '**/.*'];
export const DEFAULT_SSG_OPTIONS: SSGRunOptions = {
    displayName: 'ssg',
    triggerInstallFiles: ['package.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'],
    triggerRestartFiles: ['package.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'],
    directPaths: [],
    directRoutes: {},
    directChangeSocketOrigin: false,
    directChangeOrigin: true
};
