// @netlify/sdk/dist/constants/handler.js
export const handler = (path, config) => `import { integration } from '${path}';
integration._slug = '${config.slug}';
export const handler = integration.baseHandler;`;
export const netlifyToml = () => `[build]
command="netlify-integration build -a"
publish=".ntli/site/static"

[functions]
directory=".ntli/site/netlify/functions"
node_bundler = "esbuild"

[[headers]]
for="/ui/*"
[headers.values]
Access-Control-Allow-Origin="*"
Access-Control-Allow-Headers="*"
Access-Control-Allow-Methods="*"

[dev]
autoLaunch = false
functions = ".ntli/site/netlify/functions"
publish = ".ntli/site/static"
`;
export const integrationConfig = ({ slug = "new-integration", name = "new-integration", description, integrationLevel, parsedScopes, }) => `# yaml-language-server: $schema=https://sdk.netlify.com/sdk/config-schema.json
config:
  slug: ${slug}
  name: ${name}
  ${description ? `description: ${description}` : ""}
  ${integrationLevel ? `integrationLevel: ${integrationLevel}` : ""}
  ${parsedScopes ? parsedScopes : ""}
  # set this to true to add a global require shim to esbuild. Workaround for https://github.com/evanw/esbuild/issues/1921
  use_cjs_shims: false`;
