// @netlify/sdk/dist/cli/build/build_extension_ui.js
import { Writable } from "node:stream";
import chalk from "chalk";
import { execa } from "execa";
import { ntliLog } from "../../utils.js";
import { openExtensionDevUrl } from "../dev/open_extension_dev_url.js";
export const buildExtensionUI = async (ctx, options) => {
    let command;
    switch (options.mode) {
        case "dev":
            command = {
                file: "netlify",
                args: [
                    "dev",
                    "--no-open",
                    ...(options.previewURL ? ["--live", "integration-preview"] : []),
                ],
            };
            break;
        case "build":
            command = {
                file: "netlify",
                args: [
                    "build",
                    ...((options.siteId ?? "") === "" ? ["--offline"] : []),
                ],
            };
            break;
        default:
            throw new Error(`Unsupported Extension UI build mode: ${options.mode}`);
    }
    try {
        await execa(command.file, command.args, {
            signal: ctx.signal,
            stderr: ctx.stderr ?? undefined,
        }).pipeStdout?.(new Writable({
            write(chunk, _encoding, done) {
                const line = String(chunk);
                if (line.includes("◈ Server now ready on")) {
                    const extensionDevUrl = line.match(/◈ Server now ready on (https?:\/\/\S+?)\s/)?.[1] ??
                        "";
                    if (extensionDevUrl !== "") {
                        ntliLog(`◈ Server now ready on ${extensionDevUrl}`);
                        if (process.env.UNSTABLE__DEV_OPEN === "true") {
                            openExtensionDevUrl(extensionDevUrl, options.meta.slug);
                        }
                    }
                }
                done(null);
            },
        }));
    }
    catch (err) {
        if (err instanceof Error && "code" in err && err.code === "ENOENT") {
            console.error(chalk.yellow("Netlify CLI not found. Please install it with `npm i -g netlify-cli`"));
            process.exit(1);
        }
        throw err;
    }
};
