// @stackbit/sdk/dist/consts.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_SSG_OPTIONS = exports.EXCLUDED_COMMON_FILES = exports.EXCLUDED_DATA_FILES = exports.EXCLUDED_MARKDOWN_FILES = exports.ASSET_FILE_EXTENSIONS = exports.SUPPORTED_FILE_EXTENSIONS = exports.DATA_FILE_EXTENSIONS = exports.MARKDOWN_FILE_EXTENSIONS = void 0;
const lodash_1 = __importDefault(require("lodash"));
const excludedMarkdownFiles = ['LICENSE.md', '**/README.md', 'README.theme.md', 'CONTRIBUTING.md', 'CHANGELOG.md', 'CODE_OF_CONDUCT.md'];
exports.MARKDOWN_FILE_EXTENSIONS = ['md', 'mdx', 'markdown'];
exports.DATA_FILE_EXTENSIONS = ['yml', 'yaml', 'json', 'toml'];
exports.SUPPORTED_FILE_EXTENSIONS = [...exports.MARKDOWN_FILE_EXTENSIONS, ...exports.DATA_FILE_EXTENSIONS];
exports.ASSET_FILE_EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'svg', 'ico', 'webp', 'avif', 'tiff', 'bmp'];
exports.EXCLUDED_MARKDOWN_FILES = lodash_1.default.concat(excludedMarkdownFiles, excludedMarkdownFiles.map(lodash_1.default.toLower));
exports.EXCLUDED_DATA_FILES = ['stackbit.yaml', 'netlify.toml', 'theme.toml', '**/package.json', '**/package-lock.json', '**/yarn-lock.json'];
exports.EXCLUDED_COMMON_FILES = ['**/node_modules/**', '**/.git/**', '.idea/**', '**/.*'];
exports.DEFAULT_SSG_OPTIONS = {
    displayName: 'ssg',
    triggerInstallFiles: ['package.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'],
    triggerRestartFiles: ['package.json', 'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml'],
    directPaths: [],
    directRoutes: {},
    directChangeSocketOrigin: false,
    directChangeOrigin: true
};
