// @stackbit/sdk/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.analyzeSite = exports.matchCMS = exports.matchSSG = exports.loadContent = exports.writeConfig = exports.validateConfig = exports.mergeConfigModelsWithExternalModels = exports.loadConfigWithModelsPresetsAndValidate = exports.loadConfigFromDir = exports.loadConfig = exports.STACKBIT_CONFIG_FILES = exports.STACKBIT_CONFIG_JS_FILES = exports.STACKBIT_CONFIG_YAML_FILES = exports.mergeConfigModelsWithModelsFromFiles = exports.loadYamlModelsFromFiles = exports.getYamlModelDirs = exports.convertToYamlConfig = exports.isStackbitYamlFile = exports.findStackbitConfigFile = exports.extendModelsWithPresetsIds = exports.getPresetDirs = exports.loadPresets = void 0;
__exportStar(require("./config/config-schema"), exports);
__exportStar(require("./config/config-types"), exports);
__exportStar(require("./config/config-consts"), exports);
__exportStar(require("./content/content-errors"), exports);
__exportStar(require("./config/config-errors"), exports);
__exportStar(require("./analyzer/file-browser"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./consts"), exports);
__exportStar(require("./config/config-loader-static"), exports);
var presets_loader_1 = require("./config/presets-loader");
Object.defineProperty(exports, "loadPresets", { enumerable: true, get: function () { return presets_loader_1.loadPresets; } });
Object.defineProperty(exports, "getPresetDirs", { enumerable: true, get: function () { return presets_loader_1.getPresetDirs; } });
Object.defineProperty(exports, "extendModelsWithPresetsIds", { enumerable: true, get: function () { return presets_loader_1.extendModelsWithPresetsIds; } });
var config_loader_utils_1 = require("./config/config-loader-utils");
Object.defineProperty(exports, "findStackbitConfigFile", { enumerable: true, get: function () { return config_loader_utils_1.findStackbitConfigFile; } });
Object.defineProperty(exports, "isStackbitYamlFile", { enumerable: true, get: function () { return config_loader_utils_1.isStackbitYamlFile; } });
Object.defineProperty(exports, "convertToYamlConfig", { enumerable: true, get: function () { return config_loader_utils_1.convertToYamlConfig; } });
Object.defineProperty(exports, "getYamlModelDirs", { enumerable: true, get: function () { return config_loader_utils_1.getYamlModelDirs; } });
Object.defineProperty(exports, "loadYamlModelsFromFiles", { enumerable: true, get: function () { return config_loader_utils_1.loadYamlModelsFromFiles; } });
Object.defineProperty(exports, "mergeConfigModelsWithModelsFromFiles", { enumerable: true, get: function () { return config_loader_utils_1.mergeConfigModelsWithModelsFromFiles; } });
Object.defineProperty(exports, "STACKBIT_CONFIG_YAML_FILES", { enumerable: true, get: function () { return config_loader_utils_1.STACKBIT_CONFIG_YAML_FILES; } });
Object.defineProperty(exports, "STACKBIT_CONFIG_JS_FILES", { enumerable: true, get: function () { return config_loader_utils_1.STACKBIT_CONFIG_JS_FILES; } });
Object.defineProperty(exports, "STACKBIT_CONFIG_FILES", { enumerable: true, get: function () { return config_loader_utils_1.STACKBIT_CONFIG_FILES; } });
var config_loader_1 = require("./config/config-loader");
Object.defineProperty(exports, "loadConfig", { enumerable: true, get: function () { return config_loader_1.loadConfig; } });
Object.defineProperty(exports, "loadConfigFromDir", { enumerable: true, get: function () { return config_loader_1.loadConfigFromDir; } });
Object.defineProperty(exports, "loadConfigWithModelsPresetsAndValidate", { enumerable: true, get: function () { return config_loader_1.loadConfigWithModelsPresetsAndValidate; } });
Object.defineProperty(exports, "mergeConfigModelsWithExternalModels", { enumerable: true, get: function () { return config_loader_1.mergeConfigModelsWithExternalModels; } });
var config_validator_1 = require("./config/config-validator");
Object.defineProperty(exports, "validateConfig", { enumerable: true, get: function () { return config_validator_1.validateConfig; } });
var config_writer_1 = require("./config/config-writer");
Object.defineProperty(exports, "writeConfig", { enumerable: true, get: function () { return config_writer_1.writeConfig; } });
var content_loader_1 = require("./content/content-loader");
Object.defineProperty(exports, "loadContent", { enumerable: true, get: function () { return content_loader_1.loadContent; } });
var ssg_matcher_1 = require("./analyzer/ssg-matcher");
Object.defineProperty(exports, "matchSSG", { enumerable: true, get: function () { return ssg_matcher_1.matchSSG; } });
var cms_matcher_1 = require("./analyzer/cms-matcher");
Object.defineProperty(exports, "matchCMS", { enumerable: true, get: function () { return cms_matcher_1.matchCMS; } });
var site_analyzer_1 = require("./analyzer/site-analyzer");
Object.defineProperty(exports, "analyzeSite", { enumerable: true, get: function () { return site_analyzer_1.analyzeSite; } });
