// @sanity/schema/lib/legacy/types/crossDatasetReference.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WEAK_FIELD = exports.REF_FIELD = exports.CrossDatasetReferenceType = void 0;
var _pick2 = _interopRequireDefault(require("lodash/pick"));
var _capitalize2 = _interopRequireDefault(require("lodash/capitalize"));
var _arrify = _interopRequireDefault(require("arrify"));
var _resolve = require("../searchConfig/resolve");
var _utils = require("./utils");
var _constants = require("./constants");
var _excluded = ["name"];
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _objectWithoutProperties(e, t) { if (null == e) return {}; var o, r, i = _objectWithoutPropertiesLoose(e, t); if (Object.getOwnPropertySymbols) { var s = Object.getOwnPropertySymbols(e); for (r = 0; r < s.length; r++) o = s[r], t.includes(o) || {}.propertyIsEnumerable.call(e, o) && (i[o] = e[o]); } return i; }
function _objectWithoutPropertiesLoose(r, e) { if (null == r) return {}; var t = {}; for (var n in r) if ({}.hasOwnProperty.call(r, n)) { if (e.includes(n)) continue; t[n] = r[n]; } return t; }
var REF_FIELD = exports.REF_FIELD = {
  name: '_ref',
  title: 'Referenced document ID',
  type: 'string'
};
var WEAK_FIELD = exports.WEAK_FIELD = {
  name: '_weak',
  title: 'Weak reference marker',
  type: 'boolean'
};
var DATASET_FIELD = {
  name: '_dataset',
  title: 'The dataset containing the referenced document',
  type: 'string'
};
var REFERENCE_FIELDS = [REF_FIELD, WEAK_FIELD, DATASET_FIELD];
var OVERRIDABLE_FIELDS = [..._constants.DEFAULT_OVERRIDEABLE_FIELDS];
var CROSS_DATASET_REFERENCE_CORE = {
  name: 'crossDatasetReference',
  type: null,
  jsonType: 'object'
};
function humanize(arr, conjunction) {
  var len = arr.length;
  if (len === 1) {
    return arr[0];
  }
  var first = arr.slice(0, len - 1);
  var last = arr[len - 1];
  return "".concat(first.join(', '), " ").concat(conjunction, " ").concat(last);
}
function buildTitle(type) {
  if (!type.to || type.to.length === 0) {
    return 'Cross dataset Reference';
  }
  return "Cross dataset reference to ".concat(humanize((0, _arrify.default)(type.to).map(toType => toType.title || (0, _capitalize2.default)(toType.type)), 'or').toLowerCase());
}
var CrossDatasetReferenceType = exports.CrossDatasetReferenceType = {
  get() {
    return CROSS_DATASET_REFERENCE_CORE;
  },
  extend(subTypeDef, createMemberType) {
    if (!subTypeDef.to) {
      throw new Error("Missing \"to\" field in cross dataset reference definition. Check the type ".concat(subTypeDef.name));
    }
    var parsed = Object.assign((0, _pick2.default)(CROSS_DATASET_REFERENCE_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: CROSS_DATASET_REFERENCE_CORE
    });
    (0, _utils.lazyGetter)(parsed, 'fields', () => {
      return REFERENCE_FIELDS.map(fieldDef => {
        var name = fieldDef.name,
          type = _objectWithoutProperties(fieldDef, _excluded);
        return {
          name: name,
          type: createMemberType(type)
        };
      });
    });
    (0, _utils.lazyGetter)(parsed, 'to', () => {
      return (0, _arrify.default)(subTypeDef.to).map(toType => {
        return _objectSpread(_objectSpread({}, toType), {}, {
          // eslint-disable-next-line camelcase
          __experimental_search: (0, _resolve.resolveSearchConfigForBaseFieldPaths)(toType)
        });
      });
    });
    (0, _utils.lazyGetter)(parsed, 'title', () => subTypeDef.title || buildTitle(parsed));
    return subtype(parsed);
    function subtype(parent) {
      return {
        get() {
          return parent;
        },
        extend: extensionDef => {
          if (extensionDef.of) {
            throw new Error('Cannot override `of` of subtypes of "reference"');
          }
          var current = Object.assign({}, parent, (0, _pick2.default)(extensionDef, OVERRIDABLE_FIELDS), {
            type: parent
          });
          return subtype(current);
        }
      };
    }
  }
};
