// @sanity/schema/lib/legacy/types/text.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextType = void 0;
var _pick2 = _interopRequireDefault(require("lodash/pick"));
var _primitivePreview = _interopRequireDefault(require("../preview/primitivePreview"));
var _constants = require("./constants");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var OVERRIDABLE_FIELDS = [..._constants.DEFAULT_OVERRIDEABLE_FIELDS, 'rows'];
var TEXT_CORE = {
  name: 'text',
  type: null,
  jsonType: 'string'
};
var TextType = exports.TextType = {
  get() {
    return TEXT_CORE;
  },
  extend(subTypeDef) {
    var parsed = Object.assign((0, _pick2.default)(TEXT_CORE, OVERRIDABLE_FIELDS), subTypeDef, {
      type: TEXT_CORE,
      preview: _primitivePreview.default
    });
    return subtype(parsed);
    function subtype(parent) {
      return {
        get() {
          return parent;
        },
        extend: extensionDef => {
          var current = Object.assign({}, parent, (0, _pick2.default)(extensionDef, OVERRIDABLE_FIELDS), {
            type: parent
          });
          return subtype(current);
        }
      };
    }
  }
};
