// content-engine/dist/redux/reducers/index.js
"use strict";

exports.__esModule = true;
var _typeOwners = require("./type-owners");
exports.typeOwners = _typeOwners.typeOwnersReducer;
var _nodes = require("./nodes");
exports.nodes = _nodes.nodesReducer;
var _logs = require("../../reporter/redux/reducers/logs");
exports.logs = _logs.reducer;
var _schema = require("./schema");
exports.schema = _schema.schemaReducer;
var _definitions = require("./definitions");
exports.definitions = _definitions.definitionsReducer;
var _status = require("./status");
exports.status = _status.statusReducer;
var _config = require("./config");
exports.config = _config.configReducer;
var _lastAction = require("./last-action");
exports.lastAction = _lastAction.lastActionReducer;
var _jobsv = require("./jobsv2");
exports.jobsV2 = _jobsv.jobsV2Reducer;
var _jobs = require("./jobs");
exports.jobs = _jobs.jobsReducer;
var _nodesByType = require("./nodes-by-type");
exports.nodesByType = _nodesByType.nodesByTypeReducer;
var _program = require("./program");
exports.program = _program.programReducer;
var _resolvedNodes = require("./resolved-nodes");
exports.resolvedNodesCache = _resolvedNodes.resolvedNodesCacheReducer;
var _nodesTouched = require("./nodes-touched");
exports.nodesTouched = _nodesTouched.nodesTouchedReducer;
var _flattenedPlugins = require("./flattened-plugins");
exports.flattenedPlugins = _flattenedPlugins.flattenedPluginsReducer;
var _schemaCustomization = require("./schema-customization");
exports.schemaCustomization = _schemaCustomization.schemaCustomizationReducer;
var _inferenceMetadata = require("./inference-metadata");
exports.inferenceMetadata = _inferenceMetadata.inferenceMetadataReducer;
var _telemetry = require("./telemetry");
exports.telemetry = _telemetry.telemetryReducer;
var _setRequestHeaders = require("./set-request-headers");
exports.requestHeaders = _setRequestHeaders.setRequestHeadersReducer;
var _statefulSourcePlugins = require("./stateful-source-plugins");
exports.statefulSourcePlugins = _statefulSourcePlugins.statefulSourcePluginsReducer;
