// @stackbit/dev/dist/consts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    NEXTJS_SOURCEBIT_LIVE_UPDATE_PORT: 8088,
    DEFAULT_DIRECT_PROXY_PATH: [
        '/socket.io/**',
        '/__webpack_hmr/**',
        '/sockjs-node/**',
        '**/*.js',
        '**/*.js.map',
        '**/*.css',
        '**/*.css.map',
        '**/*.json',
        '**/*.png',
        '**/*.jpg',
        '**/*.jpeg',
        '**/*.gif',
        '**/*.svg',
        '**/*.ico',
        '**/*.pdf',
        '**/*.otf',
        '**/*.ttf',
        '**/*.tif',
        '**/*.tiff',
        '**/*.mp4',
        '**/*.mkv',
        '**/*.mov',
        '**/*.wmv',
        '**/*.zip',
        '**/*.xml',
        '**/*.ppt',
        '**/*.pptx',
        '**/*.doc',
        '**/*.docx',
        '**/*.xls',
        '**/*.xlsx',
        '**/*.rtf',
        '**/*.woff2'
    ],
    ASSET_ID_PREFIX: 'stackbit_asset_id',
    STATIC_ASSETS_FILE_PATH: '.cache/static-app-assets',
    STATIC_ASSETS_PUBLIC_PATH: '/_static/app-assets',
    STATIC_THEME_ASSETS_FILE_PATH: '.cache/static-theme-assets',
    STATIC_THEME_ASSETS_PUBLIC_PATH: '/_static/theme-assets',
    STACKBIT_URL_PATH_FIELD: 'stackbit_url_path',
    DEFAULT_CONTENT_TYPE_EXTENSIONS: ['md', 'markdown', 'mdx', 'yml', 'yaml', 'json', 'toml']
};
